wasm/names.go

```diff
@@ -0,0 +1,201 @@
+//
+// Note: The result can be nil because this does not encode empty subsections
+//
+// See https://www.w3.org/TR/wasm-core-1/#binary-namesec
+func (n *CustomNameSection) EncodeData() (data []byte) {
+	if n.ModuleName != "" {
+		data = append(data, encodeNameSubsection(uint8(0), encodeSizePrefixed([]byte(n.ModuleName)))...)
+	}
+	if fd := n.encodeFunctionNameData(); len(fd) > 0 {
+		data = append(data, encodeNameSubsection(uint8(1), fd)...)
+	}
+	if ld := n.encodeLocalNameData(); len(ld) > 0 {
+		data = append(data, encodeNameSubsection(uint8(2), ld)...)
+	}
+	return
+}
+
+// encodeFunctionNameData encodes the data for the function name subsection.
+// See https://www.w3.org/TR/wasm-core-1/#binary-funcnamesec
+func (n *CustomNameSection) encodeFunctionNameData() []byte {
+	if len(n.FunctionNames) == 0 {
+		return nil
+	}
+	return encodeSortedAndSizePrefixed(n.FunctionNames)
+}
+
+func encodeSortedAndSizePrefixed(m map[uint32]string) []byte {
+	count := uint32(len(m))
+	data := leb128.EncodeUint32(count)
+
+	// Sort the keys so that they encode in ascending order
+	keys := make(uint32Slice, 0, count)
+	for k := range m {
+		keys = append(keys, k)
+	}
+	sort.Sort(keys)
+
+	for _, i := range keys {
+		data = append(data, encodeNameMapEntry(i, []byte(m[i]))...)
+	}
+	return data
+}
+
+// encodeLocalNameData encodes the data for the local name subsection.
+// See https://www.w3.org/TR/wasm-core-1/#binary-localnamesec
+func (n *CustomNameSection) encodeLocalNameData() []byte {
+	if len(n.LocalNames) == 0 {
+		return nil
+	}
+
+	funcNameCount := uint32(len(n.LocalNames))
+	subsection := leb128.EncodeUint32(funcNameCount)
+
+	// Sort the function indices so that they encode in ascending order
+	funcIndex := make(uint32Slice, 0, funcNameCount)
+	for k := range n.LocalNames {
+		funcIndex = append(funcIndex, k)
+	}
+	sort.Sort(funcIndex)
+
+	for _, i := range funcIndex {
+		locals := encodeSortedAndSizePrefixed(n.LocalNames[i])
+		subsection = append(subsection, append(leb128.EncodeUint32(i), locals...)...)
+	}
+	return subsection
+}
+
+// encodeNameSubsection returns a buffer encoding the given subsection
+// See https://www.w3.org/TR/wasm-core-1/#subsections%E2%91%A0
+func encodeNameSubsection(subsectionID uint8, content []byte) []byte {
+	contentSizeInBytes := leb128.EncodeUint32(uint32(len(content)))
+	result := []byte{subsectionID}
+	result = append(result, contentSizeInBytes...)
+	result = append(result, content...)
+	return result
+}
+
+// encodeNameMapEntry encodes the index and data prefixed by their size.
+// See https://www.w3.org/TR/wasm-core-1/#binary-namemap
+func encodeNameMapEntry(i uint32, data []byte) []byte {
+	return append(leb128.EncodeUint32(i), encodeSizePrefixed(data)...)
+}
+
+// encodeSizePrefixed encodes the data prefixed by their size.
+func encodeSizePrefixed(data []byte) []byte {
+	size := leb128.EncodeUint32(uint32(len(data)))
+	return append(size, data...)
+}
+
+// uint32Slice implements sort.Interface
```

Comment:
I think sort.Interface is outdated one, and instead we are recommended to do
```go
var keys []uint32
// appends....
sort.Slice(keys, func (i, j int) bool { return keys[i] < keys[j] } )
```

Reply:
cool. I also found this awkward!